std/experimental/ndslice/algorithm.d

```diff
@@ -0,0 +1,1122 @@
+/**
+$(SCRIPT inhibitQuickIndex = 1;)
+
+This is a submodule of $(MREF std, experimental, ndslice).
+It contains generic multidimensional iteration algorithms.
+
+$(BOOKTABLE Cheat Sheet,
+$(TR $(TH Operator Name) $(TH Type) $(TH Functions / Seeds #) $(TH Tensors #) $(TH Returns) $(TH First Argument))
+$(T6 ndMap, Lazy, `>=1`, `1`, Tensor, Tensor)
+$(T6 ndFold, Eagerly, `>=1`, `1`, Scalar, Tensor)
+$(T6 ndReduce, Eagerly, `1`, `>=1`, Scalar, Seed)
+$(T6 ndEach, Eagerly, `0`, `>=1`, `void`, Tensor)
+)
+
+All operators are suitable to change tensors using `ref` argument qualification in a function declaration.
+
+Note:
+    $(SUBREF iteration, transposed) and
+    $(SUBREF selection, pack) can be used to specify dimensions.
+
+License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors:   Ilya Yaroshenko
+
+Source:    $(PHOBOSSRC std/_experimental/_ndslice/_algorithm.d)
+
+Macros:
+SUBREF = $(REF_ALTTEXT $(TT $2), $2, std,experimental, ndslice, $1)$(NBSP)
+T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
+T6=$(TR $(TDNW $(LREF $1)) $(TD $2) $(TD $3) $(TD $4) $(TD $5) $(TD $6))
+*/
+module std.experimental.ndslice.algorithm;
+
+import std.traits;
+import std.meta;
+
+import std.experimental.ndslice.internal;
+import std.experimental.ndslice.slice;
+
+private template TensorFronts(size_t length)
+{
+    static if (length)
+    {
+        enum i = length - 1;
+        enum TensorFronts = TensorFronts!(length - 1) ~ "tensors[" ~ i.stringof ~ "].front, ";
+    }
+    else
+    {
+        enum TensorFronts = "";
+    }
+}
+
+private void checkTensors(bool seed, Args...)(auto ref Args tensors)
+{
+    enum msg = seed ?
+        "all arguments except the first (seed or needle) must be tensors" :
+        "all arguments must be tensors"
+        ~ tailErrorMessage!();
+    enum msgShape = "all tensors must have the same shape"  ~ tailErrorMessage!();
+    foreach (i, Arg; Args)
+    {
+        static assert (is(Arg == Slice!(N, Range), size_t N, Range), msg);
+        static if (i)
+        {
+            static assert (tensors[i].N == tensors[0].N, msgShape);
+            assert(tensors[i].shape == tensors[0].shape, msgShape);
+        }
+    }
+}
+
+private template naryFun(bool seed, size_t argsLengs, alias fun)
+{
+    static if (argsLengs + seed == 1)
+    {
+        import std.functional : unaryFun;
+        alias naryFun = unaryFun!fun;
+    }
+    else
+    static if (argsLengs + seed == 2)
+    {
+        import std.functional : binaryFun;
+        alias naryFun = binaryFun!fun;
+    }
+    else
+    {
+        alias naryFun = fun;
+    }
+}
+
+/++
+Implements the homonym function (also known as `transform` present
+in many languages of functional flavor. The call `ndMap!(fun)(tensor)`
```

Comment:
Closing parenthesis was added.
Other modules in `ndslice` has not proper docs too. I am not good in documentation writng